UM/Settings/ContainerQuery.py

```diff
@@ -56,50 +60,50 @@ def getResult(self) -> Optional[List[Dict[str, Any]]]:
+        key_so_far = (self._ignore_case, )  # type: Tuple[Any, ...]
         if candidates is None:
-            candidates = list(self._registry.metadata.values())
-        filtered_candidates = cast(Iterable, candidates)
+            filtered_candidates = list(self._registry.metadata.values())
+        else:
+            filtered_candidates = candidates
 
         # Filter on all the key-word arguments.
         for key, value in self._kwargs.items():
+            key_so_far += (key, value)
+            if candidates is None and key_so_far in self.cache:
+                filtered_candidates = cast(List[Dict[str, Any]], self.cache[key_so_far].getResult())
+                self.cache.move_to_end(key_so_far)  # The cache entry was used, so make sure it doesn't fall off as an infrequently-used cache entry.
+                continue
+
             if isinstance(value, type):
-                key_filter = functools.partial(self._matchType, property_name=key, value=value)
+                key_filter = functools.partial(self._matchType, property_name = key, value = value)
             elif isinstance(value, str):
-                # It's a string.
                 if ContainerQuery.OPTIONS_REGEX.fullmatch(value) is not None:
-                    # With [token1|token2|token3|...], we try to find if any of the given tokens is present in the value
+                    # With [token1|token2|token3|...], we try to find if any of the given tokens is present in the value.
                     key_filter = functools.partial(self._matchRegMultipleTokens, property_name = key, value = value)
                 elif ("*" or "|") in value:
                     key_filter = functools.partial(self._matchRegExp, property_name = key, value = value)
                 else:
-                    key_filter = functools.partial(self._matchString, property_name = key, value=value)
+                    key_filter = functools.partial(self._matchString, property_name = key, value = value)
             else:
-                key_filter = functools.partial(self._matchDirect, property_name=key, value=value)
-            filtered_candidates = filter(key_filter, filtered_candidates)
-
-        # Execute all filters.
-        self._result = list(filtered_candidates)
+                key_filter = functools.partial(self._matchDirect, property_name = key, value = value)
+            filtered_candidates = list(filter(key_filter, filtered_candidates))  # Execute this filter.
 
-    def __hash__(self) -> int:
-        return hash(self.__key())
+            if candidates is None:  # Only cache if we didn't pre-filter candidates.
+                cached_arguments = dict(zip(key_so_far[1::2], key_so_far[2::2]))
+                self.cache[key_so_far] = ContainerQuery(self._registry, ignore_case = self._ignore_case, **cached_arguments)  # Cache this query for the next time.
+                self.cache[key_so_far]._result = filtered_candidates
```

Comment:
The container query didn't give consent to its privates being touched.

Reply:
Similar to above: Normally this is allowed since it's the same class. It's allowed in Java and C++ at least.